cdc/sink/codec/canal_flat.go

```diff
@@ -159,49 +217,85 @@ func (c *CanalFlatEventBatchEncoder) newFlatMessageForDML(e *model.RowChangedEve
 
 	// We need to ensure that both Data and Old have exactly one element,
 	// even if the element could be nil. Changing this could break Alibaba's adapter
-	ret.Data = append(ret.Data, data)
-	ret.Old = append(ret.Old, oldData)
+	flatMessage.Data = append(flatMessage.Data, data)
+	flatMessage.Old = append(flatMessage.Old, oldData)
+
+	if !c.enableTiDBExtension {
+		return flatMessage, nil
+	}
 
-	return ret, nil
+	return &canalFlatMessageWithTiDBExtension{
+		canalFlatMessage: flatMessage,
+		Extensions:       &tidbExtension{CommitTs: e.CommitTs},
+	}, nil
 }
 
-func (c *CanalFlatEventBatchEncoder) newFlatMessageForDDL(e *model.DDLEvent) *canalFlatMessage {
+func (c *CanalFlatEventBatchEncoder) newFlatMessageForDDL(e *model.DDLEvent) canalFlatMessageInterface {
 	header := c.builder.buildHeader(e.CommitTs, e.TableInfo.Schema, e.TableInfo.Table, convertDdlEventType(e), 1)
-
-	ret := &canalFlatMessage{
+	flatMessage := &canalFlatMessage{
 		ID:            0, // ignored by both Canal Adapter and Flink
 		Schema:        header.SchemaName,
 		Table:         header.TableName,
 		IsDDL:         true,
 		EventType:     header.GetEventType().String(),
 		ExecutionTime: header.ExecuteTime,
-		BuildTime:     0, // ignored by both Canal Adapter and Flink
+		BuildTime:     time.Now().UnixNano() / 1e6, // timestamp
 		Query:         e.Query,
 		tikvTs:        e.CommitTs,
 	}
-	return ret
+
+	if !c.enableTiDBExtension {
+		return flatMessage
+	}
+
+	return &canalFlatMessageWithTiDBExtension{
+		canalFlatMessage: flatMessage,
+		Extensions:       &tidbExtension{CommitTs: e.CommitTs},
+	}
+}
+
+func (c *CanalFlatEventBatchEncoder) newFlatMessage4CheckpointEvent(ts uint64) *canalFlatMessageWithTiDBExtension {
+	return &canalFlatMessageWithTiDBExtension{
+		canalFlatMessage: &canalFlatMessage{
+			ID:            0,
+			IsDDL:         false,
+			EventType:     tidbWaterMarkType,
+			ExecutionTime: convertToCanalTs(ts),
+			BuildTime:     time.Now().UnixNano() / 1e6, // converts to milliseconds
```

Comment:
```suggestion
			BuildTime:     time.Now().UnixNano() / int64(time.Millisecond), // converts to milliseconds
```
can update too for other places using `1e6`.